### PR TITLE
[NVBug 6108145] Fix PTQ calibration and export for fused-experts MoE (Qwen3.5-MoE VLM)

### PR DESCRIPTION
**Pull request by meenchen**
### What does this PR do?

Type of change: Bug fix

  Fixes a 4-bug cascade that caused silent PTQ failure on Qwen3.5-MoE VLMs (Qwen3.6-35B-A3B): calibration
  appeared to succeed but produced token-salad at inference. Root cause: HF's @use_experts_implementation
  dispatches expert forward to torch._grouped_mm / torch.bmm, bypassing the F.linear hook that captures
  activations — so gate_up_proj_input_quantizer / down_proj_input_quantizer never calibrated and no input_scale
  tensors were emitted.

  Changes:
  - examples/llm_ptq/hf_ptq.py — force config._experts_implementation = "eager" (recursing into text_config /
  vision_config / …) so per-expert F.linear calls are visible to the calibration hook.
  - modelopt/torch/quantization/conversion.py — normalize plural ModuleList quantizer names (weight_quantizers.N
  → weight_quantizer) before fnmatch, so wildcards like *mlp.experts*weight_quantizer match fused-expert
  quantizers.
  - modelopt/torch/export/unified_export_hf.py — hoist the _QuantFusedExperts export branch above the
  get_quantization_format() gate so _export_fused_experts() runs even when the top-level format query returns
  QUANTIZATION_NONE (happens for experts-only recipes).
  - modelopt_recipes/general/ptq/nvfp4_experts_only-fp8_kv.yaml — layerwise: false (VLM nested layer structure
  breaks the layerwise walker).

### Usage

```python
  python examples/llm_ptq/hf_ptq.py \
      --pyt_ckpt_path Qwen/Qwen3.6-35B-A3B \
      --qformat nvfp4 \
      --kv_cache_qformat fp8 \
      --calib_size 512 \
      --export_path Qwen3.6-35B-A3B-NVFP4
```

### Testing

  Testing

  End-to-end PTQ → vLLM deploy → NEL eval on Qwen3.6-35B-A3B (256 experts × 40 layers, 35B params):

   Hook-call diagnostic: 0 → 6720 per-expert F.linear calls during calibration after the fix; 0 → 30720
  input_scale tensors emitted in the exported checkpoint.

  FP8 fused-MoE path still produces gibberish — separate follow-up (vLLM per-expert weight_scale handling).
*  vLLM full-FP8: the FlashInfer TRTLLM Fp8MoE loader doesn't stack the 256 per-expert scalar weight_scale tensors
   into a [num_experts] per-expert vector — it ends up applying one expert's scale across all 256, so every
  routed expert dequants with the wrong amplitude → coherent token stream collapses into multilingual gibberish.

 * SGLang full-FP8: qwen3_5.py::_make_packed_weight_loader rejects with AssertionError: Unexpected scalar for
  tuple shard load: loaded_shard_id=(0,1,2), split_sizes=[1,1,1] — its packed-loader has no path for "N
  independent per-tensor source scalars combining into one fused-shard parameter," so the fused QKV (or
  in_proj_qkvz) load is structurally refused and the model never finishes loading.
### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / ❌ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * Better fused-expert export flow, a plugin to force eager expert execution during calibration/export, and a representative quantizer discovery utility.

* **Bug Fixes**
  * Reliable matching/discovery of per-expert indexed quantizers enabling correct calibration and mixed-precision export; fixes for calibration in nested decoder layouts.

* **Documentation**
  * Clarified PTQ config guidance on layerwise calibration.

* **Tests**
  * Added fused-experts calibration, export, and name-normalization tests.

### CODE REVIEW

File: modelopt/torch/export/unified_export_hf.py [hunk 1]
```diff
@@ -646,6 +646,15 @@ def _process_quantized_modules(
...
         ):
             sub_module.unpack_weight()
+        # _QuantFusedExperts uses plural `gate_up_proj_weight_quantizers` (ModuleList),
+        # which get_quantization_format's singular-weight_quantizer check misses. Handle
+        # it explicitly before the format gate so fused-experts get split + quantized.
+        if hasattr(sub_module, "gate_up_proj_weight_quantizers"):
+            from modelopt.torch.export.moe_utils import _export_fused_experts
+
+            with fsdp2_aware_weight_update(model, sub_module, reshard=False):
+                _export_fused_experts(sub_module, dtype)
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py:658]: This new early-exit block makes the existing `elif hasattr(sub_module, "gate_up_proj_weight_quantizers")` block at the end of this function (inside the `get_quantization_format(sub_module) != QUANTIZATION_NONE` branch, around line ~700 in the full file) dead code — the `continue` here means we never reach that path.

Please remove the old dead-code block to avoid confusion for future maintainers who might not realize both paths exist.

**Review by cjluo-nv — COMMENTED** [review]
This is a well-structured bug fix with good test coverage. The 4-part fix (eager impl forcing, quantizer name normalization, export path hoisting, YAML layerwise change) is logically coherent and well-tested. However, there's a duplicate code path in unified_export_hf.py that should be cleaned up.

**Design**: Despite the complexity gate firing, this is a bug fix within existing systems, not an architectural change. No new abstractions are introduced.

**Tests**: Comprehensive — covers `force_eager_experts_impl_on_the_fly` edge cases, and an end-to-end calibration test that guards the full pipeline (name normalization → wildcard matching → amax collection). Good.

**Issue**: The new early-exit block in `_process_quantized_modules` makes the existing `elif hasattr(sub_module, "gate_up_proj_weight_quantizers")` block (deeper in the same function, inside the `get_quantization_format() != QUANTIZATION_NONE` guard) dead code. One of these should be removed.

File: modelopt/torch/export/unified_export_hf.py [hunk 2]
```diff
@@ -646,6 +646,15 @@ def _process_quantized_modules(
             "QuantFP8Linear" in type(sub_module).__name__ and sub_module.weight.element_size() <= 1
         ):
             sub_module.unpack_weight()
+        # _QuantFusedExperts uses plural `gate_up_proj_weight_quantizers` (ModuleList),
+        # which get_quantization_format's singular-weight_quantizer check misses. Handle
+        # it explicitly before the format gate so fused-experts get split + quantized.
+        if hasattr(sub_module, "gate_up_proj_weight_quantizers"):
+            from modelopt.torch.export.moe_utils import _export_fused_experts
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py]: move import to the top

File: modelopt/torch/export/unified_export_hf.py [hunk 3]
```diff
@@ -646,6 +646,15 @@ def _process_quantized_modules(
...
+
+            with fsdp2_aware_weight_update(model, sub_module, reshard=False):
+                _export_fused_experts(sub_module, dtype)
+            continue
```
Comment (cjluo-nv) [inline modelopt/torch/export/unified_export_hf.py]: should this be else instead for readability?

File: modelopt/torch/quantization/conversion.py [hunk 4]
```diff
@@ -286,6 +287,26 @@ def set_quantizer_by_cfg(quant_model: nn.Module, quant_cfg: QuantizeQuantCfgType
...
+    ``nn.ModuleList``s named e.g. ``gate_up_proj_weight_quantizers`` — its children
+    get dotted names like ``...gate_up_proj_weight_quantizers.0``. These don't match
+    the singular-suffix wildcards (``*weight_quantizer``) used in the stock configs,
+    so the experts stay at their defaults. Return a normalized name where
+    ``weight_quantizers.N`` / ``input_quantizers.N`` collapse to their singular form
+    so the standard wildcards match.
+    """
+    return _FUSED_EXPERTS_QUANTIZER_LIST_RE.sub(
+        lambda m: m.group(1)[:-1],
+        name,  # "weight_quantizers" -> "weight_quantizer"
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/conversion.py]: do we want to cover the case like ...gate_up_proj_weight_quantizer.0 if the quantizer naming is not a plural form?

File: modelopt/torch/export/unified_export_hf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (meenchen) [reply modelopt/torch/export/unified_export_hf.py]: Done in c47a6022 — removed the unreachable elif block. The early-exit `continue` at the new `elif` handles all fused-experts modules before the `get_quantization_format()` gate, so the later branch is no longer reachable.

File: modelopt/torch/export/unified_export_hf.py (hunk 1 above)
Comment (meenchen) [reply modelopt/torch/export/unified_export_hf.py:658]: Done in c47a6022 — the dead `elif hasattr(sub_module, "gate_up_proj_weight_quantizers")` block inside the `get_quantization_format() != QUANTIZATION_NONE` branch has been removed. Only the early-exit at the top of the loop remains.

File: modelopt/torch/export/unified_export_hf.py (hunk 2 above)
Comment (meenchen) [reply modelopt/torch/export/unified_export_hf.py]: Done in c47a6022 — `_export_fused_experts` now imported at the top of the file (alphabetized between `model_utils` and `plugins`).

File: modelopt/torch/export/unified_export_hf.py (hunk 3 above)
Comment (meenchen) [reply modelopt/torch/export/unified_export_hf.py]: Done in c47a6022 — restructured as `elif` chained to the preceding `unpack_weight()` branch. The fused-experts path is now visibly one of the disjoint options at the top of the loop, with `continue` to skip the `get_quantization_format()` gate below.

File: modelopt/torch/quantization/conversion.py (hunk 4 above)
Comment (meenchen) [reply modelopt/torch/quantization/conversion.py]: Good call — broadened in c47a6022. The regex is now `(weight_quantizers?|input_quantizers?)\.\d+(?=$|\.)` and the substitution conditionally strips the trailing `s` only when present, so it handles both:

- `...gate_up_proj_weight_quantizers.0` → `...gate_up_proj_weight_quantizer` (plural — current `_QuantFusedExperts`)
- `...gate_up_proj_weight_quantizer.0`  → `...gate_up_proj_weight_quantizer` (singular — defensive for future variants)

Added `TestNormalizeFusedExpertsQuantizerName` covering plural, singular, non-indexed passthrough, and unrelated `.N` patterns.

File: modelopt/torch/export/unified_export_hf.py (hunk 3 above)
Comment (meenchen) [reply modelopt/torch/export/unified_export_hf.py]: Good catch — the prior commit chained `unpack_weight` and the fused-experts branch as `if/elif`, which papered over the structure (unpack is preprocessing that needs to fall through; only the export branches should be disjoint). Restructured in 5a11aeb2:

```python
# Preprocessing — falls through
if hasattr(sub_module, "weight_packed") or QuantFP8Linear...:
    sub_module.unpack_weight()

# Two disjoint export paths
if hasattr(sub_module, "gate_up_proj_weight_quantizers"):
    # fused experts
    ...
elif get_quantization_format(sub_module) != QUANTIZATION_NONE:
    # standard path
    ...
```

The fused-experts `continue` is gone; `elif` makes the two export paths mutually exclusive without needing the explicit short-circuit.

**Review by cjluo-nv — COMMENTED** [review]
All critical previous review comments have been addressed: dead code removed, import moved to top, control flow restructured, and singular/plural quantizer naming covered. The PR is a well-structured 4-part bug fix (eager impl forcing, quantizer name normalization, export path hoisting, YAML layerwise change) with comprehensive tests. No new abstractions — purely surgical fixes within existing systems. Code is correct and clean.

Complex PR: spans 6 directories (≥ 5). Looping in a human for approval.

**Review by cjluo-nv — APPROVED** [review]

**Review by sychen52 — APPROVED** [review]
